connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java

```diff
@@ -215,7 +215,12 @@ protected <V> V execAndRetry(ProcessingContext<T> context, Operation<V> operatio
                 } else {
                     log.trace("Can't retry. start={}, attempt={}, deadline={}", startTime, attempt, deadline);
                     context.error(e);
-                    return null;
+                    markAsFailed();
+                    if (withinToleranceLimits()) {
+                        return null;
+                    } else {
+                        throw new ConnectException("Exceeded deadline & tolerance for retriable exception", e);
+                    }
                 }
                 if (stopping) {
                     log.trace("Shutdown has been scheduled. Marking operation as failed.");
```

Comment:
This data loss scenario is making me think about this `stopping` flag, and whether it could cause data loss. Running out of retries and stopping retries due to a shutdown should probably behave similarly, and probably shouldn't skip the record.

But stopping is only set by `#triggerStop`/`WorkerTask#cancel`, which is a hard-shutdown operation after no further data is expected from the task, and no offsets are being committed.

I think we can probably leave this in-place, and if we ever change task shutdown we can address it then.

Reply:
Agreed. In both source/sinks the worker doesn't commit the offsets if it's been cancelled. So data loss should not be possible because of that conditional check. Will leave as is.